### DISCUSSION TITLE
cy.within() - subject not updated or assertion simply not retried as mentioned in docs?

### DISCUSSION THREAD

**Discussion by Cleudi** [cypress-io/cypress#14994] [Questions and Help]
Hi, 
since I am not sure if this is a bug, desired behavior or just some mistake in our code, I prefer to ask before opening a bug:

The flow: 

- we fill an input field, 
- a request is sent to the server (suggest-like) and 
- if nothing is received, an error message is displayed. 
- We check if the error message is displayed correctly. Yes, we see it on screen, but cypress does not recognize it - until we add wait(1000) before querying the field to check!

Details:
- we have a custom command for this input field, which wraps the subject in `within()`, then queries for the error 
- as soon as we add `wait(1000)` before calling the command, the error message is recognized by cypress. Basically this means that we get the server response before we call the command.
- if we call the code for the error query outside of the custom command/within, the error message is received as well. 

In cypress docs I found the following information: [assertions are not retried in within](https://docs.cypress.io/api/commands/within.html#Assertions). Is this the root cause for our issue? And if yes: is this expected behavior? Are there recommended strategies how to deal with that, eg. `cy.intercept`(listening for the response) before calling the custom command or an alternative pattern without using within?

Thanks for ideas 😄 

----------------- Pseudocode ---------------

No `wait`, no `within` -> **works**
```
      cy.get('mybutton')
        .click() // request to server is fired
        .get('mySuggest')
        .get('myErrorIcon').should('have.length',1);
```

`wait`, `within` -> **works**
```
      cy.get('mybutton')
        .click() // request to server is fired
        .wait(1000)
        .get('mySuggest')
        .within((subject) => {
          cy.wrap(subject).get('myErrorIcon').should('have.length', 1);
      });
```

No `wait`, `within` -> **does not work**
```
      cy.get('mybutton')
        .click() // request to server is fired
        .get('mySuggest')
        .within((subject) => {
          cy.wrap(subject).get('myErrorIcon').should('have.length', 1);
      });
```

**Comment by bahmutov** [comment]
Can you please provide a minimal reproducible example, since it is hard to guess what is going on from just a description

**Reply by Cleudi** [reply]
I added pseudocode for all three examples...

**Reply by bahmutov** [reply]
We cannot run the pseudo code to see what your web application is doing in this case. Does the subject element itself change after the request? By wrapping it and by adding `.get('myErrorIcon')` you "lock in" the subject and Cypress only retries the last command before the assertion, https://on.cypress.io/retry-ability

I would write the above test simpler to use a combined selector

```js
cy.get('mybutton')
  .click() // request to server is fired
cy.get('my Subject myErrorIcon').should('have.length', 1);
```
